src/components/MultiSelect/MultiSelect.tsx

```diff
@@ -1,138 +1,479 @@
-  onClear?: () => void;
-  /** The type of MultiSelect that will be rendered. */
-  type: "listbox" | "dialog";
+  /** listOverflow is a property indicating how the list should handle overflow, with options limited to either "scroll" or "expand." */
+  listOverflow?: MultiSelectListOverflowTypes;
   /** The selected items state (items that were checked by user). */
   selectedItems: SelectedItems;
   /** Value used to set the width for the MultiSelect component. */
   width?: MultiSelectWidths;
-}
-type ListboxOnChange = (selectedItem: MultiSelectItem, id: string) => void;
-type DialogOnChange = (event: React.ChangeEvent<HTMLInputElement>) => void;
-
-type MultiSelectVariantsProps =
-  | {
-      type: "listbox";
-      onApply?: never;
-      /** The action to perform for downshift's onSelectedItemChange function. */
-      onChange: ListboxOnChange;
-      // These are props that are never allowed on the listbox type.
-      onMixedStateChange?: never;
-    }
-  | {
-      type: "dialog";
-      /** The action to perform for save/apply button of multiselect. */
-      onApply: () => void;
-      /** The action to perform on the checkbox's onChange function.  */
-      onChange: DialogOnChange;
-      /** The action to perform for a mixed state checkbox (parent checkbox). */
-      onMixedStateChange?: DialogOnChange;
-    };
-
-export type MultiSelectProps = MultiSelectCommonProps &
-  MultiSelectVariantsProps;
+};
 
 /**
- * The `MultiSelect` component is a form input element that presents a list
- * of `Checkbox` components from which a user can make one or multiple
- * selections. Two variants of the MultiSelect component are offered, each with
- * slightly different functionality and requirements.  Because of these
- * differences, the two variants are broken out in separate stories below.
- */
+  The MultiSelect component is a customizable form input that supports multiple configurations,
+  including search functionality, checkbox options, and hierarchical structure,
+  with a parent checkbox toggling all children and dynamic styling through Chakra UI.
+*/
+
 export const MultiSelect = chakra(
   forwardRef<HTMLDivElement, React.PropsWithChildren<MultiSelectProps>>(
-    (props, ref?) => {
+    (props, _ref?) => {
       const {
+        helperText,
         id,
         isBlockElement = false,
         isDefaultOpen = false,
+        isSearchable = false,
+        defaultItemsVisible = 5,
         items,
-        labelText,
-        onApply,
+        listOverflow = "scroll",
+        buttonText,
         onChange,
         onClear,
         onMixedStateChange,
         selectedItems,
-        type,
-        width = "default",
+        width = "full",
         ...rest
       } = props;
 
-      const commonProps = {
-        id,
+       // Create a ref that we add to the element for which we want to detect outside clicks.
+       const internalRef: React.RefObject<HTMLDivElement> =
```

Comment:
This can also use a better variable name because even though it is an internal ref, the way that it is used is very specific for the focus behavior.